cns/deviceplugin/server.go

```diff
@@ -0,0 +1,157 @@
+package deviceplugin
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"time"
+
+	"github.com/pkg/errors"
+	"go.uber.org/zap"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+	"k8s.io/kubelet/pkg/apis/deviceplugin/v1beta1"
+)
+
+const devicePrefix = "NIC-"
+
+type deviceCounter interface {
+	getDeviceCount() int
+}
+
+type Server struct {
+	address                  string
+	logger                   *zap.Logger
+	deviceCounter            deviceCounter
+	listAndWatchShutdownChan chan struct{}
+	deviceCheckInterval      time.Duration
+}
+
+func NewServer(logger *zap.Logger, address string, deviceCounter deviceCounter, deviceCheckInterval time.Duration) *Server {
+	return &Server{
+		address:                  address,
+		logger:                   logger,
+		deviceCounter:            deviceCounter,
+		listAndWatchShutdownChan: make(chan struct{}),
+		deviceCheckInterval:      deviceCheckInterval,
+	}
+}
+
+// Run starts the grpc server and blocks until an error or context is cancelled. Wait on Ready to know when the server is ready.
+func (s *Server) Run(ctx context.Context) error {
+	grpcServer := grpc.NewServer()
+	v1beta1.RegisterDevicePluginServer(grpcServer, s)
+	defer close(s.listAndWatchShutdownChan)
```

Comment:
instead of closing the shutdownChan ourselves, we could use the context that we get here to simplify further:
```suggestion
func (s *Server) Run(ctx context.Context) error {
	grpcServer := grpc.NewServer()
	v1beta1.RegisterDevicePluginServer(grpcServer, s)
	
	childCtx, cancel := context.WithCancel(ctx)
	defer cancel()
	s.shutdownCh = childCtx.Done() // assign the ctx Done signal to the shutdown ch so we don't have to close anything explicitly
```

Reply:
Hmm. Updated